### PR TITLE
global: start removing g_ceph_context

### PR DESCRIPTION
**Pull request by adamemerson**
Unskew the clock,
Add a `dout_context` macro to simplify handling `dout` without rewriting too much code
Remove uses of g_ceph_context from the OSD, ObjectStore, and object class code.

More to come!

Stay tuned kids for our next exciting commit.
Same squid time
Same squid repo!

🦑

### CODE REVIEW

**Comment by liewegas** [comment]
:+1: Let's rebase this on master and merge it quick before it grows more conflicts

**Comment by liewegas** [comment]
Passed rados test, except for the clock skew test (which I didn't remove yet).  @adamemerson any update on the replacement test?

Also there are a couple trivial conflicts that a rebase would resolve

**Comment by adamemerson** [comment]
Just got back from Oregon so reintegrating myself into everything. I'll rebase this and deal with Loic's whitespace concern and get an updated test in shortly.

**Comment by adamemerson** [comment]
@dachary I got rid of the relevant white space stuff (though I think we should clean some of it up and some point,(

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -1,4 +1,3 @@
-// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: keep this!

File: src/include/encoding.h [hunk 2]
```diff
@@ -24,10 +24,61 @@
...
+class CephContext;
+
 using namespace ceph;
 
+namespace ceph {
+  // Occasionally an object requiring a CephContext* will need to be
+  // decoded. This object can be default constructed in every field
+  // /except/ the CephContext*. To avoid contagion (that is, to avoid
+  // having to shove a CephContext* into *every* object to preserve a
+  // uniform interface, make a specialized template function to help.
```
Comment (liewegas) [inline src/include/encoding.h]: I don't see any users for this yet... ?

Comment (liewegas) [inline src/include/encoding.h]: Should we leave it out until there are users?

**Comment by liewegas** [comment]
Again, let's minimize scope so we can test and merge before we get new conflicts

**Comment by adamemerson** [comment]
Sure, I can pull out the encode change. I want to get the *rest* in, though, since it keeps moving away from me and then I have to rebase and retest.

File: src/os/bluestore/BlueStore.cc (hunk 1 above)
Comment (adamemerson) [inline src/os/bluestore/BlueStore.cc]: Oops, sorry, that was unintentional. I'll fix that.

File: src/include/encoding.h (hunk 2 above)
Comment (adamemerson) [inline src/include/encoding.h]: We can leave it out. I just threw it in here to give people a chance to look at it and see if it was wrong-headed. The first user is PGMap inside the monitor. I'll hold it until I get that done. I just wanted to get what I had working up before master skittered away from me.

**Comment by liewegas** [comment]
np.  looks reasonable, but it'll help to see the user.  usually the 
encode/decode stuff is applied to data types that don't need cct

**Comment by liewegas** [comment]
@athanatos does this look ok to you?

**Pull request merged**